src/Components/Endpoints/src/Builder/RazorComponentEndpointDataSource.cs

```diff
@@ -166,4 +171,34 @@ public override IChangeToken GetChangeToken()
         Debug.Assert(_endpoints != null);
         return _changeToken;
     }
+
+    // Test only method to allow tracking change token dipose calls in unit tests.
+    internal void EnableChangeTokenDisposeTracking()
+    {
+        _enableChangeTokenDisposeTracking = true;
+    }
+
+    public class WrappedChangeTokenDisposable : IDisposable
+    {
+        public bool IsDisposed { get; private set;}
+        private readonly IDisposable _innerDisposable;
+
+        public WrappedChangeTokenDisposable(IDisposable innerDisposable)
+        { 
+            _innerDisposable = innerDisposable;
+        }
+
+        public void Dispose()
+        { 
+             IsDisposed = true;
+             _innerDisposable.Dispose();
+        }
+    }
```

Comment:
This this type is only used for testing, could we find a way to move it to the test project?

An alternative that might not require as many new members could be adding a field of type `Func<IDisposable>` to this class, and it acts as the factory for creating the `_disposableChangeToken`. The field's value after construction is just a delegate that does what currently happens [here](https://github.com/dotnet/aspnetcore/blob/26f9f27c8fbbdddd75d1b9ac34715cf45bf1578b/src/Components/Endpoints/src/Builder/RazorComponentEndpointDataSource.cs#L149). Then the test could assign a new factory (either directly or via `internal` property) that wraps the default one and does all the tracking it needs to.